crates/agent/src/publications/status.rs

```diff
@@ -0,0 +1,193 @@
+    },
+    TestFailed,
+    PublishFailed,
+    Success {
+        /// If any materializations are to be updated in response to this publication,
+        /// their publication ids will be included here. This is purely informational.
+        #[serde(default, skip_serializing_if = "Vec::is_empty")]
+        linked_materialization_publications: Vec<Id>,
+    },
+    /// Returned when there are no draft specs (after pruning unbound
+    /// collections). There will not be any `draft_errors` in this case, because
+    /// there's no `catalog_name` to associate with an error. And it may not be
+    /// desirable to treat this as an error, depending on the scenario.
+    EmptyDraft,
+    /// One or more expected `last_pub_id`s did not match the actual `last_pub_id`, indicating that specs
+    /// have been changed since the draft was created.
+    ExpectPubIdMismatch {
+        #[serde(default, skip_serializing_if = "Vec::is_empty")]
+        failures: Vec<LockFailure>,
+    },
+}
+
+impl JobStatus {
+    pub fn is_success(&self) -> bool {
+        // TODO(phil): should EmptyDraft also be considered successful?
+        // This question is not relevent today, but will become important
+        // once we implement auto-discovery in controllers.
+        match self {
+            JobStatus::Success { .. } => true,
+            _ => false,
+        }
+    }
+
+    pub fn has_incompatible_collections(&self) -> bool {
+        matches!(self, JobStatus::BuildFailed { incompatible_collections, .. } if !incompatible_collections.is_empty())
+    }
+
+    pub fn is_empty_draft(&self) -> bool {
+        matches!(self, JobStatus::EmptyDraft)
+    }
+
+    pub fn success() -> JobStatus {
+        JobStatus::Success {
+            // This field is no longer used, but is being kept around for a while in case we need
+            // parse an existing status that includes it.
+            linked_materialization_publications: Vec::new(),
+        }
+    }
+    pub fn build_failed(incompatible_collections: Vec<IncompatibleCollection>) -> JobStatus {
+        JobStatus::BuildFailed {
+            incompatible_collections,
+            evolution_id: None,
+        }
+    }
+}
+
+/// Represents an optimistic lock failure when trying to update live specs.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct LockFailure {
+    pub catalog_name: String,
+    pub expect_pub_id: models::Id,
+    pub last_pub_id: Option<models::Id>,
+}
+
+/// Reasons why a draft collection spec would need to be published under a new name.
+#[derive(Serialize, Deserialize, Clone, Copy, PartialEq, Debug)]
+#[serde(rename_all = "camelCase")]
+pub enum ReCreateReason {
+    /// The collection key in the draft differs from that of the live spec.
+    KeyChange,
+    /// One or more collection partition fields in the draft differs from that of the live spec.
+    PartitionChange,
+    /// A live spec with the same name has already been created and was subsequently deleted.
```

Comment:
nit: remove?

Reply:
I'd been leaving all the existing publication status fields, just in case we needed to deserialize an old status for some reason. That's now seeming overly cautious, though, so I just removed this and the `linked_materializations` stuff.